lib/vector-core/src/stream/partitioned_batcher.rs

```diff
@@ -405,40 +296,45 @@ where
                         this.closed_batches.extend(
                             this.batches
                                 .drain()
-                                .map(|(key, batch)| (key, batch.into_inner())),
+                                .map(|(key, mut batch)| (key, batch.take_batch())),
                         );
                         continue;
                     }
                     return Poll::Ready(None);
                 }
                 Poll::Ready(Some(item)) => {
                     let item_key = this.partitioner.partition(&item);
-                    let item_limit: usize = *this.batch_item_limit;
-                    let alloc_limit: usize = *this.batch_allocation_limit;
-
-                    if let Some(batch) = this.batches.get_mut(&item_key) {
-                        if batch.has_space(&item) {
-                            // When there's space in the partition batch just
-                            // push the item in and loop back around.
-                            batch.push(item);
-                        } else {
-                            let new_batch = Batch::new(item_limit, alloc_limit).with(item);
-                            let batch = mem::replace(batch, new_batch);
-
-                            // The batch for this partition key was set to
-                            // expire, but now it's overflowed and must be
-                            // pushed out, so now we reset the batch timeout.
-                            this.timer.insert(item_key.clone());
 
-                            this.closed_batches.push((item_key, batch.into_inner()));
-                        }
+                    // Get the batch for this partition, or create a new one.
+                    let batch = if let Some(batch) = this.batches.get_mut(&item_key) {
+                        batch
                     } else {
-                        // We have no batch yet for this partition key, so
-                        // create one and create the expiration entries as well.
-                        // This allows the batch to expire before filling up,
-                        // and vice versa.
-                        let batch = Batch::new(item_limit, alloc_limit).with(item);
+                        let batch = (this.state)();
                         this.batches.insert(item_key.clone(), batch);
+                        this.batches
+                            .get_mut(&item_key)
+                            .expect("batch has just been inserted so should exist")
+                    };
```

Comment:
Would using `let batch = this.batches.get_mut(&item_key).unwrap_or_else(|| …);` simplify this any?

Reply:
If I recall I ran into lifetime issues, but I'll give it another go.

Reply:
You mean [this](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.get_or_insert_with)?

Reply:
Ah, possibly. Wrong container... :-(